### ISSUE TITLE
Question regarding Parameter Regularization in ToMoE Paper

### ISSUE THREAD

**Issue by leeglg** [gaosh/ToMoE#1]
Thank you for releasing ToMoE paper.

I have a question regarding the calculation of the R_P loss described in Section 3.4 Parameter Regularization. 

While the calculation method for the active width of MLP layers (d^MLP_l) is clearly presented, I have a confusion for how the active width of MHA layers (d^MHA_l) is determined.

How can i use s_0, s_t to calculate R_P? Have you choose max among s_t of tokens in the batch?

**Comment by gaosh** [comment]
Yeah, s_t is chosen from the maximum among the tokens. s_0 is static pruning, so it is fixed for all tokens.

**Comment by leeglg** [comment]
thanks for answering!

**Comment by leeglg** [comment]
Hello again, the paper is very interesting, so I'm trying to reproduce it. I have some questions.


The size of the value of in figure 3. (c), R_U value scale is difficult to understand.

In equations (6) and (7), I think **\Sigma{u_l} / | u_l | ** represents a ratio,  setting features not selected by any masks to 0.

So, The function   f ( \Sigma{u_l} / | u_l |  , 1 )  is always the log value of 1 / (used feature ratio). 

The smaller the ratio, the larger the value, and as the ratio approaches 1, the value approaches 0.

If so, the behavior of R_U  in (c), starting from 0, approaching 1, and then converging to around 0.1,  
    is understood as initially using all features (when R_U=0),
    then using only about 10% of the total features (when R_U ≈1), 
    and finally converging to using about 80% of the total features (when R_U ≈0.1).


![Image](https://github.com/user-attachments/assets/f0e576a2-0a10-4410-8d73-fb001e09ed45)

However, considering the pruning ratio p presented in (c), I understand that during the learning process, masks become excessively sparse, exceeding the target pruning rate(0.4, 0.5, 0.6, and then they converge to the pruning ratio or to some another value.


Could you please tell me if this analysis/observation is correct?



As I understand it, R_U  in (c) is  R_U_MLP + R_U_MHA (i dont know it has to be  mean or sum.)

In my reproduce, R_U is always close to 0, and never exceeds 0.003,  even at its maximum. 
I think I might have implemented it wrong somewhere.

![Image](https://github.com/user-attachments/assets/a0f3feb0-a743-4d6f-ad28-792d47f9a9d7)

**Comment by gaosh** [comment]
Your analysis is right. For R_U, it should go up first and gradually go down to a value close to 0. If you further extend the number of iterations, it will become even closer to zero. I think the learning dynamics are not correct in the figure you presented. I provided a simplified code sample for the R_U implementation. I hope it helps.

```python
#minmax regularization loss
def minmax_reg_loss(x,y,c=0):
    x=x+c
    y=y+c
    max_tensor = torch.maximum(x,y)
    min_tensor = torch.minimum(x,y)

    loss = torch.log(max_tensor/min_tensor)
    return loss.mean()

def experts_union(experts_feature):
    return 1 - torch.prod(1 - experts_feature, dim=0)

union_of_experts = experts_union(binary)
r_u = minmax_reg_loss(union_of_experts.mean(), torch.scalar_tensor(1).to(device).float())
```

**Comment by leeglg** [comment]
Thank you for your specific and clear answer. It helped me a lot to solve the problem with my code. Is it included as s_0 in calculating the union loss? Right now I'm only using s_e (from MLP), s_t (from MHA) masks in calculating the union loss.

> 
    # 2-1. Parameter loss MLP
    loss_param_mlp = get_regularizer_param_mlp(model)
    # 2-2. Parameter loss MHA s_0
    loss_param_mha_s0 = get_regularizer_param_mha_s0(model)
    # 2-3. Parameter loss MHA s_t
    loss_param_mha_st = get_regularizer_param_mha_st(st_masks_list, 
                                                     attention_mask_idx)
    loss_param = cal_loss_param(model, llama_config, tomoe_config, loss_param_mlp, loss_param_mha_s0, loss_param_mha_st)
    
    # 2-4. Union loss MLP
    loss_union_mlp = get_regularizer_union_mlp(model)
    # 2-5. Union loss MHA s_t
    loss_union_mha = get_regularizer_union_mha_st(st_masks_list, 
                                               attention_mask_idx)
    loss_union = loss_union_mlp + loss_union_mha
    
    # 2-6. Loadbalance loss MLP
    loss_loadbal = get_regularizer_loadbalance(routerlogits_list, 
                                               onehotg_list, 
                                               num_actual_token, 
                                               attention_mask_idx)
    reg_loss = (tomoe_config.alpha * loss_param) + (tomoe_config.beta * loss_union) + (tomoe_config.gamma * loss_loadbal)

**Comment by gaosh** [comment]
s_0 is not used to calculated the union loss since it is static. In addition, the calculation of the loss_param is done globally, I usally implemented it by calcualting the number of remained parameters from all layers (MLP & MHA), and calculate the loss with the remained parameters and the number of total parameters.
